src/lib-smtp/fuzz-smtp-server.c

```diff
@@ -0,0 +1,96 @@
+/* Copyright (c) 2020 Catena cyber */
+
+#include "lib.h"
+#include "str.h"
+#include "array.h"
+#include "hostpid.h"
+#include "ioloop.h"
+#include "istream.h"
+#include "ostream.h"
+#include "time-util.h"
+#include "sleep.h"
+#include "connection.h"
+#include "test-common.h"
+#include "smtp-address.h"
+#include "smtp-reply-parser.h"
+#include "smtp-server.h"
+
+#include <sys/types.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <fcntl.h>
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size);
+
+static struct ioloop *ioloop_fuzz;
+static int initialized = 0;
+int fdo[2];
+int fdi[2];
+
+struct server_connection {
+    void *context;
+};
+
+
+static void server_connection_destroy(void *context)
+{
+    struct server_connection *sconn = (struct server_connection *)context;
+    i_free(sconn);
+}
+
+static void
+test_server_defaults(struct smtp_server_settings *smtp_set)
+{
+    /* server settings */
+    i_zero(smtp_set);
+    smtp_set->max_client_idle_time_msecs = 500;
+    smtp_set->max_pipelined_commands = 16;
+    smtp_set->auth_optional = TRUE;
+    smtp_set->debug = 0;
+}
+
+
+static void test_server_continue(void *dummy) {
+    //instead of simple io_loop_stop so as to free input io
+    io_loop_stop_delayed(ioloop_fuzz);
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+    struct smtp_server_connection *conn;
+    struct server_connection *sconn;
+    struct timeout *to;
+    struct smtp_server_settings smtp_server_set;
+    static struct smtp_server_callbacks server_callbacks;
+    static struct smtp_server *smtp_server = NULL;
+
+    if (initialized == 0) {
+        lib_init();
+        hostpid_init();
+        initialized = 1;
+    }
```

Comment:
This looks like something that should be put to FUZZ_START_* macros automatically (hostpid_init() is called internally by lib_init() already):

```
diff --git a/src/lib-test/fuzzer.h b/src/lib-test/fuzzer.h
index b0c3d7625a..24ec60998b 100644
--- a/src/lib-test/fuzzer.h
+++ b/src/lib-test/fuzzer.h
@@ -1,13 +1,18 @@
 #ifndef FUZZER_H
 #define FUZZER_H
 
+#define FUZZ_BEGIN_COMMON \
+	if (!lib_is_initialized) lib_init();
+
 #define FUZZ_BEGIN_DATA(args) \
 	int LLVMFuzzerTestOneInput(args); \
-	int LLVMFuzzerTestOneInput(args) {
+	int LLVMFuzzerTestOneInput(args) { \
+		FUZZ_BEGIN_COMMON T_BEGIN {
 
 #define FUZZ_BEGIN_STR(str_arg) \
 	int LLVMFuzzerTestOneInput(const uint8_t *_param_data, size_t _param_size); \
 	int LLVMFuzzerTestOneInput(const uint8_t *_param_data, size_t _param_size) { \
+		FUZZ_BEGIN_COMMON \
 		T_BEGIN { str_arg = t_strndup(_param_data, _param_size);
 
 #define FUZZ_END \
```

Reply:
I added this to https://github.com/dovecot/core/pull/121